### PR TITLE
Stack Refactor

### PR DESCRIPTION
**Pull request by haldenl**
Refactor stack to inside X / Y encoding channels. Issue #343

### CODE REVIEW

**Comment by kanitw** [comment]
- [] For `isStack` perhaps, implement it so that 
  1. It checks if all required properties are _not_ wildcards.
  2. generate a spec from the specQuery -- you may need to modify the `toSpec/toEncoding` to still return a spec/encoding even though some unnecessary properties are still wildcards
  3. Once you get `mark`, `encoding`, `config` -- you can call VL's `stack` method.

**Comment by haldenl** [comment]
Should users be able to specify a 'stack' wildcard in a non-XY channel (in which case only 'null' stack would pass constraints). To reword, should anything be returned if stack wildcard is specified in say, the 'color' channel?

**Comment by kanitw** [comment]
> Should users be able to specify a 'stack' wildcard in a non-XY channel (in which case only 'null' stack would pass constraints). To reword, should anything be returned if stack wildcard is specified in say, the 'color' channel?


I think it's fine -- since null is one of the valid values.  But users specify a wildcard that enumerate specific values without null, then it should only work with x and y.

**Comment by kanitw** [comment]
From our conversation, two cases to prevent:

- If a non-null stacked is assigned to non-xy (e.g., color), we can add an encoding (field) constraint to prevent this -- this should be a _strict_ constraint. 

- if stacked is specified, but the plot setup is compatible with stacking (Vega-Lite stack method will return no stack) -- in this case we can use vega-lite stack method to double check.

**Comment by haldenl** [comment]
From conversation:

If stack is implicitly specced, then we should ignore constraints for stack, let VL take care of it.

**Comment by kanitw** [comment]
I think if you rebase correctly, there shouldn't be this many commimts.

**Comment by haldenl** [comment]
oh i haven't squashed yet, isn't that what collapses the commits?

File: src/query/spec.ts [hunk 1]
```diff
@@ -127,97 +128,99 @@ export function isAggregate(specQ: SpecQuery) {
...
 
-  if (stackBy.length === 0) {
-    return null;
+/**
+ * Returns true iff the given property does not contain a nested wildcard.
+ * @param prop The property in question.
+ * @param opt With optional `exclude` property, which defines properties to
+ * ignore when testing for wildcards.
+ */
+function containsWildcard(prop: any, opt: {exclude?: {[key: string]: 1}} = {}) {
```
Comment (kanitw) [inline src/query/spec.ts]: The function and parameter name do not indicate what you describe in the comment very well. 

Consider renaming the method and use generic?

File: src/query/spec.ts [hunk 2]
```diff
@@ -127,97 +128,99 @@ export function isAggregate(specQ: SpecQuery) {
...
+  if (prop !== Object(prop)) {
```
Comment (kanitw) [inline src/query/spec.ts]: we have `isObject` in util

File: test/constraint/spec.test.ts [hunk 3]
```diff
@@ -926,45 +954,119 @@ describe('constraints/spec', () => {
...
             ]
           });
-          assert.isTrue(SPEC_CONSTRAINT_INDEX['omitNonSumStack'].satisfy(specM, schema, DEFAULT_QUERY_CONFIG));
+
+          assert.isTrue(SPEC_CONSTRAINT_INDEX['omitInvalidStackSpec'].satisfy(specM, schema, DEFAULT_QUERY_CONFIG));
+        });
+      });
+    });
+
+    it('should return false for stack specified in incorrect channel', () => {
```
[the remark this answers is not in the document]
Comment (haldenl) [reply test/constraint/spec.test.ts]: TODO: add test for channel: '?', stack: '?'

File: src/query/shorthand.ts [hunk 4]
```diff
@@ -96,31 +98,21 @@ export function spec(specQ: SpecQuery,
...
-      const field = fieldDef(_stack.fieldEncQ, includeExceptType, replace);
-      const groupby = _stack.groupByEncQ ? fieldDef(_stack.groupByEncQ, includeExceptType, replace) : undefined;
-
-      parts.push(
-        'stack={field:' + field + ',' +
-        (groupby ? 'by:' + groupby + ',' : '') +
-        'offset:' + _stack.offset + '}'
-      );
-    }
+    stack = getVlStack(specQ);
```
[the remark this answers is not in the document]
Comment (haldenl) [reply src/query/shorthand.ts]: we use the stack.fieldChannel as well!

File: src/query/spec.ts [hunk 5]
```diff
@@ -127,97 +128,99 @@ export function isAggregate(specQ: SpecQuery) {
...
+export function getVlStack(specQ: SpecQuery): StackProperties {
+  if (!hasRequiredStackProperties(specQ)) {
     return null;
   }
 
-  // Should have stackable mark
-  if (!contains(['bar', 'area'], specQ.mark)) {
-    return null;
+  const encoding = toEncoding(specQ.encodings, {schema: null, wildcardMode: 'null'});
+  const mark = specQ.mark as Mark;
```
[the remark this answers is not in the document]
Comment (haldenl) [reply src/query/spec.ts]: How would I go about this? The check at the top of the method should ensure specQ.mark is not a wildcard if that is the concern.

File: test/constraint/spec.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (haldenl) [reply test/constraint/spec.test.ts]: Actually, I think the channel: '?', stack: '?' case should be covered by the test in field.test.ts

File: test/constraint/spec.test.ts [hunk 6]
```diff
@@ -926,45 +954,119 @@ describe('constraints/spec', () => {
...
+        assert.isTrue(SPEC_CONSTRAINT_INDEX['omitInvalidStackSpec'].satisfy(specM, schema, DEFAULT_QUERY_CONFIG));
       });
     });
-  });
 
-  describe('omitNonSumStack', () => {
-    it('should return true if summative-based aggregate is used.', () => {
-      SUM_OPS.forEach((aggregate) => {
-        [Channel.OPACITY, Channel.DETAIL, Channel.COLOR].forEach((stackByChannel) => {
+    it('should return true if stack defined without wildcard', () => {
```
[the remark this answers is not in the document]
Comment (haldenl) [reply test/constraint/spec.test.ts]: done!

File: test/constraint/spec.test.ts [hunk 7]
```diff
@@ -870,53 +870,81 @@ describe('constraints/spec', () => {
...
-    it('should return true if color uses a non-linear scale when it is mapped to a non-X or non-Y channel that is aggregate', () => {
-      const specM = buildSpecQueryModel({
-        mark: Mark.BAR,
-        encodings: [
-          {channel: Channel.X, field: 'A', type: Type.QUANTITATIVE, aggregate: 'sum'},
-          {channel: Channel.Y, field: 'B', type: Type.NOMINAL},
-          {channel: Channel.COLOR, field: 'C', type: Type.QUANTITATIVE, scale: {type: ScaleType.POW}},
-          {channel: Channel.DETAIL, field: 'A', type: Type.NOMINAL}
-        ]
+    it('should return true for stack with linear scale type', () => {
```
[the remark this answers is not in the document]
Comment (haldenl) [reply test/constraint/spec.test.ts]: oops, fixed.

File: src/query/spec.ts (hunk 2 above)
Comment (haldenl) [reply src/query/spec.ts]: done!

File: src/query/spec.ts [hunk 8]
```diff
@@ -127,97 +128,99 @@ export function isAggregate(specQ: SpecQuery) {
...
-        channel: encQ.channel,
-        fieldDef: encQ
-      });
+/**
+ * Returns true iff the given SpecQuery has the properties defined
+ * to be a potential Stack spec.
+ * @param specQ The SpecQuery in question.
+ */
+export function hasRequiredStackProperties(specQ: SpecQuery) {
+  const requiredProps = [specQ.mark, specQ.encodings.filter(encQ => !isDisabledAutoCountQuery(encQ))];
```
[the remark this answers is not in the document]
Comment (haldenl) [reply src/query/spec.ts]: discussion prior

File: src/query/spec.ts [hunk 9]
```diff
@@ -127,97 +128,99 @@ export function isAggregate(specQ: SpecQuery) {
...
+  return stack(mark, encoding, getStackOffset(specQ));
+}
+
+/**
+ * @return The `StackOffset` specified in `specQ`, `undefined` if none
+ * is specified.
+ */
+export function getStackOffset(specQ: SpecQuery): StackOffset {
+  for (const encQ of specQ.encodings) {
+    if (typeof encQ[Property.STACK] !== 'undefined' && !isWildcard(encQ[Property.STACK])) {
```
[the remark this answers is not in the document]
Comment (haldenl) [reply src/query/spec.ts]: done

File: src/constraint/spec.ts [hunk 10]
```diff
@@ -692,40 +688,40 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
...
+      }
+
       return true;
     }
   },
   {
-    name: 'omitNonSumStack',
-    description: 'Stacked plot should use summative aggregation such as sum, count, or distinct',
-    properties: [Property.CHANNEL, Property.MARK, Property.AGGREGATE, Property.AUTOCOUNT],
+    name: 'omitNonSummativeStackSpec',
```
[the remark this answers is not in the document]
Comment (haldenl) [reply src/constraint/spec.ts]: reverted!

File: src/constraint/spec.ts [hunk 11]
```diff
@@ -692,40 +688,40 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
...
-             contains([Channel.X, Channel.Y], encQ.channel)) {
-              if (scaleType(encQ) !== ScaleType.LINEAR) {
-                return false;
-            }
-          }
-        }
+      if (!specM.wildcardIndex.hasProperty(Property.STACK)) {
+        return true;
       }
+      const specStack = specM.getVlStack();
```
[the remark this answers is not in the document]
Comment (haldenl) [reply src/constraint/spec.ts]: done

File: src/constraint/field.ts [hunk 12]
```diff
@@ -339,6 +339,19 @@ export const FIELD_CONSTRAINTS: EncodingConstraintModel<FieldQuery>[] = [
       }
       return true;
     }
+  },
+  {
+    name: 'stackMatchesChannel',
```
[the remark this answers is not in the document]
Comment (haldenl) [reply src/constraint/field.ts]: done

File: src/query/spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (haldenl) [reply src/query/spec.ts]: Also to only check for the required encoding properties, I define an array of required properties:

` const requiredEncodingProps = [Property.STACK, Property.CHANNEL,
      Property.MARK, Property.FIELD, Property.AGGREGATE, Property.AUTOCOUNT, Property.SCALE,
      getEncodingNestedProp('scale', 'type'), Property.TYPE];`

and then `without(ALL_ENCODING_PROPS, requredEncodingProps)` to give to `exclude`.

File: src/constraint/spec.ts [hunk 13]
```diff
@@ -692,40 +688,40 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
...
+      }
+
       return true;
     }
   },
   {
     name: 'omitNonSumStack',
-    description: 'Stacked plot should use summative aggregation such as sum, count, or distinct',
-    properties: [Property.CHANNEL, Property.MARK, Property.AGGREGATE, Property.AUTOCOUNT],
+    description: 'Stack specifications that use non-summative aggregates should be omitted (even implicit ones)',
```
[the remark this answers is not in the document]
Comment (haldenl) [reply src/constraint/spec.ts]: problematic spec if implicit is not enforced

{
  "spec": {
    "data": {"url": "data/cars.json"},
    "mark": "bar",
    "encodings": [
      {
        "channel": "?",
        "field": "Cylinders",
        "type": "nominal"
      },
      {
        "channel": "?",
        "field": "Origin",
        "type": "ordinal"
      },
      {
        "channel": "?",
        "aggregate": "mean",
        "field": "Acceleration",
        "type": "quantitative"
      }
    ]
  },
  "orderBy": "effectiveness"
}

[the remark this answers is not in the document]
Comment (haldenl) [reply src/constraint/spec.ts]: result:

image

[the remark this answers is not in the document]
Comment (haldenl) [reply src/constraint/spec.ts]: with implicit constrained:

image

File: src/constraint/spec.ts [hunk 14]
```diff
@@ -692,40 +688,40 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
...
-              if (scaleType(encQ) !== ScaleType.LINEAR) {
-                return false;
-            }
-          }
-        }
+      if (!specM.wildcardIndex.hasProperty(Property.STACK)) {
+        return true;
       }
+      const stackProps = specM.getVlStack();
+      if (stackProps == null) {
```
[the remark this answers is not in the document]
Comment (haldenl) [reply src/constraint/spec.ts]: done

File: src/constraint/spec.ts [hunk 15]
```diff
@@ -692,40 +688,40 @@ export const SPEC_CONSTRAINTS: SpecConstraintModel[] = [
...
     allowWildcardForProperties: false,
-    strict: false,
+    strict: true,
     satisfy: (specM: SpecQueryModel, _: Schema, __: QueryConfig) => {
-      const stack = specM.stack();
-      if (stack) {
-        const measureEncQ = specM.getEncodingQueryByChannel(stack.fieldChannel);
-        return (isFieldQuery(measureEncQ) &&  contains(SUM_OPS, measureEncQ.aggregate)) || (isAutoCountQuery(measureEncQ) && !!measureEncQ.autoCount);
+      const specStack = specM.getVlStack();
+      if (specStack != null) {
```
[the remark this answers is not in the document]
Comment (haldenl) [reply src/constraint/spec.ts]: done

File: src/constraint/spec.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (haldenl) [reply src/constraint/spec.ts]: discussed -- we need this constraint

**Pull request merged**